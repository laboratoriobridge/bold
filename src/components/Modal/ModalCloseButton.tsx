import React from 'react'
import { useLocale } from '../../i18n'
import { ExternalStyles } from '../../styles'
import { Button } from '../Button'
import { Icon } from '../Icon'
import { Tooltip } from '../Tooltip'

interface ModalCloseButton {
  onClose: () => void
  style?: ExternalStyles
}

export const ModalCloseButton = (props: ModalCloseButton) => {
  const { onClose, style: externalStyles } = props

  const locale = useLocale()

  return (
    <Tooltip text={locale.modal.close}>
      <Button aria-label={locale.modal.close} size='small' skin='ghost' onClick={onClose} style={externalStyles}>
        <Icon icon='timesDefault' />
      </Button>
    </Tooltip>
  )
}
