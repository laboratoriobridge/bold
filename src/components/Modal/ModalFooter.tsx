import React, { CSSProperties } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'

export interface ModalFooterProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  style?: ExternalStyles
}

export function ModalFooter(props: ModalFooterProps) {
  const { children, style, ...rest } = props

  const { classes, css } = useStyles(createStyles)

  return (
    <div className={css(classes.footer, style)} {...rest}>
      {children}
    </div>
  )
}

const createStyles = (theme: Theme) => ({
  footer: {
    backgroundColor: theme.pallete.surface.background,
    padding: '1rem',
    borderBottomLeftRadius: theme.radius.modal,
    borderBottomRightRadius: theme.radius.modal,
    height: '5rem',
    width: '100%',
    borderTop: `1px solid ${theme.pallete.divider}`,
    gridColumn: '1 / -1',
    gridRow: '3',
  } as CSSProperties,
})
