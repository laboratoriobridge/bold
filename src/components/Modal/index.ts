export * from './auto'
export { Modal, ModalProps, ModalSize, ModalScroll } from './Modal'
export { ModalBackdrop, ModalBackdropProps } from './ModalBackdrop'
export { ModalHeader, ModalHeaderChildrenProps, ModalHeaderContentProps } from './ModalHeader'
export { ModalBody, ModalBodyProps } from './ModalBody'
export { ModalContainer, ModalContainerProps } from './ModalContainer'
export { ModalFooter, ModalFooterProps } from './ModalFooter'
export { ModalFooterButton, ModalFooterButtonProps } from './ModalFooterButton'
export { ModalHeaderIconObject, ModalHeaderIconType } from './ModalHeaderIcon'
export { ModalCloseButton, ModalCloseButtonProps } from './ModalCloseButton'
export { ModalSidebar, ModalSidebarProps, ModalSidebarPosition } from './ModalSidebar'
