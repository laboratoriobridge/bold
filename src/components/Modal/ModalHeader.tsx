import React, { CSSProperties, ReactNode, useEffect } from 'react'
import { Theme, useStyles } from '../../styles'
import { Heading } from '../Heading'
import { HFlow } from '../HFlow'
import { VFlow } from '../VFlow'
import { useIsOverflowing } from '../../hooks'
import { useModalContext } from '../../hooks'
import { ModalCloseButton } from './ModalCloseButton'
import { ModalHeaderIconType, ModalHeaderIcon } from './ModalHeaderIcon'

type ModalHeaderBaseProps = {
  hasCloseButton?: boolean
  onCloseButtonClick?: () => void
}

export type ModalHeaderContentProps = ModalHeaderBaseProps & {
  title: ReactNode
  subtitle?: ReactNode
  icon?: ModalHeaderIconType
}

export type ModalHeaderChildrenProps = ModalHeaderBaseProps & {
  children: ReactNode
}

export function ModalHeader(props: ModalHeaderContentProps): JSX.Element
export function ModalHeader(props: ModalHeaderChildrenProps): JSX.Element

export function ModalHeader(props: ModalHeaderContentProps | ModalHeaderChildrenProps) {
  const { hasCloseButton = true, onCloseButtonClick } = props

  const { scroll, bodyRef, setHasHeader } = useModalContext()
  const isBodyOverflowing = useIsOverflowing(bodyRef, 'vertical')
  const { classes } = useStyles(createStyles, scroll === 'body' && isBodyOverflowing)

  useEffect(() => {
    setHasHeader(true)
    return () => setHasHeader(false)
  }, [setHasHeader])

  return (
    <HFlow
      hSpacing={0.5}
      justifyContent='space-between'
      alignItems='flex-start'
      style={classes.header}
      data-testid='modal-header'
    >
      {isHeaderWithChildren(props) ? (
        props.children
      ) : (
        <HFlow hSpacing={1} justifyContent='flex-start' alignItems='center'>
          {props.icon && <ModalHeaderIcon icon={props.icon} />}
          <VFlow vSpacing={0}>
            <Heading level={1} color='normal' fontWeight='bold'>
              {props.title}
            </Heading>
            {props.subtitle && (
              <Heading level={5} color='normal' fontWeight='normal'>
                {props.subtitle}
              </Heading>
            )}
          </VFlow>
        </HFlow>
      )}
      {hasCloseButton && <ModalCloseButton onClick={onCloseButtonClick} />}
    </HFlow>
  )
}

const isHeaderWithChildren = (
  props: ModalHeaderContentProps | ModalHeaderChildrenProps
): props is ModalHeaderChildrenProps => 'children' in props

const createStyles = (theme: Theme, showHeaderShadow: boolean) => ({
  header: {
    width: '100%',
    gridTemplateColumns: '1fr auto',
    padding: '1.5rem 1rem 1rem 2rem',
    zIndex: 1,
    backgroundColor: theme.pallete.surface.main,
    boxShadow: showHeaderShadow ? theme.shadows.outer[10] : '',
  } as CSSProperties,
})
