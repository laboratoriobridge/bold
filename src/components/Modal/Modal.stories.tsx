import { action } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs'
import React from 'react'
import { TextColor } from '../../styles'
import { textColorMap } from '../../styles/theme/createPallete'
import { Button } from '../Button'
import { Icons } from '../Icon'
import { IconMap } from '../Icon/generated/types'
import { VFlow } from '../VFlow'

import { HFlow } from '../HFlow'
import { Switch } from '../Switch'
import { modal } from './auto'
import { ModalMountTarget } from './auto/ModalMountTarget'
import { Modal, ModalScroll, ModalSize } from './Modal'
import { ModalBody } from './ModalBody'
import { ModalFooter } from './ModalFooter'
import { ModalFooterButton } from './ModalFooterButton'

const sizes: { [key in ModalSize]: ModalSize } = {
  large: 'large',
  small: 'small',
  auto: 'auto',
}

const scrolls: { [key in ModalScroll]: ModalScroll } = {
  body: 'body',
  full: 'full',
}

const icons: Icons[] = Object.keys(IconMap) as Icons[]
const iconColors: TextColor[] = Object.keys(textColorMap) as TextColor[]

export default {
  title: 'Components/Modal',
}

export const Default = () => {
  const open = boolean('open', true)
  const size = select('size', sizes, 'large')
  const scroll = select('scroll', scrolls, 'body')
  const title = text('title', 'Modal Title')
  const subtitle = text('subtitle', 'Modal Subtitle')
  const hasIcon = boolean('hasIcon', true)
  const icon = hasIcon ? select('icon', icons, 'bridge') : undefined
  const iconFill = hasIcon ? select('iconFill', iconColors, 'normal') : undefined
  const hasCloseIcon = boolean('hasCloseIcon', true)
  const manageOverflow = boolean('manageOverflow', true)
  const closeOnBackdropClick = boolean('closeOnBackdropClick', true)
  const onClose = action('close')

  return (
    <div>
      <p>Default</p>

      <Modal
        open={open}
        size={size}
        scroll={scroll}
        hasCloseIcon={hasCloseIcon}
        title={title}
        subtitle={subtitle}
        icon={icon}
        iconFill={iconFill}
        manageOverflow={manageOverflow}
        closeOnBackdropClick={closeOnBackdropClick}
        onClose={onClose}
      >
        <ModalBody>
          <VFlow>
            <p>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer molestie, dui id luctus commodo, nunc
              enim facilisis odio, vel hendrerit erat neque eu nisl. Donec urna felis, pharetra nec urna a, efficitur
              lobortis urna. Mauris varius purus vehicula lorem mollis, a cursus enim malesuada. Integer at congue enim.
              Nullam purus mauris, fermentum nec mattis in, cursus nec tellus. Nunc sodales orci tortor, at feugiat
              purus hendrerit a. Suspendisse potenti. Nam porta urna vitae nibh pharetra eleifend. Nullam urna eros,
              auctor vitae maximus non, feugiat eget odio. Cras venenatis, lectus eget consectetur volutpat, urna felis
              efficitur enim, vitae viverra purus risus sed purus.
            </p>

            <p>
              Aliquam placerat eget massa vel congue. Suspendisse libero erat, auctor ac tristique et, ultrices quis
              dolor. Suspendisse commodo lacinia lectus ut egestas. Cras in urna eget leo mollis luctus eu eget felis.
              Mauris vel libero vel lectus cursus placerat. Nunc non ligula urna. Aliquam id quam at quam sodales
              gravida condimentum quis eros. am dolor nisl, ullamcorper ac varius nec, vulputate sed metus. Cras lacus
              ante, dapibus facilisis risus at, varius commodo nulla. Nunc volutpat pellentesque mauris et sodales.
              Suspendisse finibus, dui at tristique mollis, felis mauris vulputate leo, sed vehicula lacus nibh ut
              dolor.
            </p>
            <p>
              Aliquam placerat eget massa vel congue. Suspendisse libero erat, auctor ac tristique et, ultrices quis
              dolor. Suspendisse commodo lacinia lectus ut egestas. Cras in urna eget leo mollis luctus eu eget felis.
              Mauris vel libero vel lectus cursus placerat. Nunc non ligula urna. Aliquam id quam at quam sodales
              gravida condimentum quis eros. am dolor nisl, ullamcorper ac varius nec, vulputate sed metus. Cras lacus
              ante, dapibus facilisis risus at, varius commodo nulla. Nunc volutpat pellentesque mauris et sodales.
              Suspendisse finibus, dui at tristique mollis, felis mauris vulputate leo, sed vehicula lacus nibh ut
              dolor.
            </p>
            <p>
              Aliquam placerat eget massa vel congue. Suspendisse libero erat, auctor ac tristique et, ultrices quis
              dolor. Suspendisse commodo lacinia lectus ut egestas. Cras in urna eget leo mollis luctus eu eget felis.
              Mauris vel libero vel lectus cursus placerat. Nunc non ligula urna. Aliquam id quam at quam sodales
              gravida condimentum quis eros. am dolor nisl, ullamcorper ac varius nec, vulputate sed metus. Cras lacus
              ante, dapibus facilisis risus at, varius commodo nulla. Nunc volutpat pellentesque mauris et sodales.
              Suspendisse finibus, dui at tristique mollis, felis mauris vulputate leo, sed vehicula lacus nibh ut
              dolor.
            </p>
            <p>
              Aliquam placerat eget massa vel congue. Suspendisse libero erat, auctor ac tristique et, ultrices quis
              dolor. Suspendisse commodo lacinia lectus ut egestas. Cras in urna eget leo mollis luctus eu eget felis.
              Mauris vel libero vel lectus cursus placerat. Nunc non ligula urna. Aliquam id quam at quam sodales
              gravida condimentum quis eros. am dolor nisl, ullamcorper ac varius nec, vulputate sed metus. Cras lacus
              ante, dapibus facilisis risus at, varius commodo nulla. Nunc volutpat pellentesque mauris et sodales.
              Suspendisse finibus, dui at tristique mollis, felis mauris vulputate leo, sed vehicula lacus nibh ut
              dolor.
            </p>
            <p>
              Aliquam placerat eget massa vel congue. Suspendisse libero erat, auctor ac tristique et, ultrices quis
              dolor. Suspendisse commodo lacinia lectus ut egestas. Cras in urna eget leo mollis luctus eu eget felis.
              Mauris vel libero vel lectus cursus placerat. Nunc non ligula urna. Aliquam id quam at quam sodales
              gravida condimentum quis eros. am dolor nisl, ullamcorper ac varius nec, vulputate sed metus. Cras lacus
              ante, dapibus facilisis risus at, varius commodo nulla. Nunc volutpat pellentesque mauris et sodales.
              Suspendisse finibus, dui at tristique mollis, felis mauris vulputate leo, sed vehicula lacus nibh ut
              dolor.
            </p>
            <p>
              Aliquam placerat eget massa vel congue. Suspendisse libero erat, auctor ac tristique et, ultrices quis
              dolor. Suspendisse commodo lacinia lectus ut egestas. Cras in urna eget leo mollis luctus eu eget felis.
              Mauris vel libero vel lectus cursus placerat. Nunc non ligula urna. Aliquam id quam at quam sodales
              gravida condimentum quis eros. am dolor nisl, ullamcorper ac varius nec, vulputate sed metus. Cras lacus
              ante, dapibus facilisis risus at, varius commodo nulla. Nunc volutpat pellentesque mauris et sodales.
              Suspendisse finibus, dui at tristique mollis, felis mauris vulputate leo, sed vehicula lacus nibh ut
              dolor.
            </p>
          </VFlow>
        </ModalBody>
        <ModalFooter>
          <HFlow alignItems='center' justifyContent='flex-end'>
            <ModalFooterButton onClick={action('cancel clicked')}>Cancel</ModalFooterButton>
            <ModalFooterButton kind='primary' onClick={action('save clicked')}>
              Save
            </ModalFooterButton>
          </HFlow>
        </ModalFooter>
      </Modal>
    </div>
  )
}

export const Auto = () => (
  <div>
    <Button
      onClick={modal({
        title: 'Confirm',
        size: 'small',
        render: () => 'Are you sure?',
        actions: [
          { label: 'Cancel', onClick: action('Cancel') },
          { label: 'Ok', kind: 'primary', onClick: action('Ok') },
        ],
      })}
    >
      Auto modal
    </Button>
    <ModalMountTarget />
  </div>
)

export const ModalOverlap = () => (
  <div>
    <p>Modal Overlap</p>

    <Modal
      title='Modal Overlap'
      open={boolean('open', true)}
      size={select('main modal size', sizes, 'large')}
      onClose={action('close')}
    >
      <ModalBody>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer molestie, dui id luctus commodo, nunc enim
          facilisis odio, vel hendrerit erat neque eu nisl. Donec urna felis, pharetra nec urna a, efficitur lobortis
          urna. Mauris varius purus vehicula lorem mollis, a cursus enim malesuada. Integer at congue enim. Nullam purus
          mauris, fermentum nec mattis in, cursus nec tellus. Nunc sodales orci tortor, at feugiat purus hendrerit a.
          Suspendisse potenti. Nam porta urna vitae nibh pharetra eleifend. Nullam urna eros, auctor vitae maximus non,
          feugiat eget odio. Cras venenatis, lectus eget consectetur volutpat, urna felis efficitur enim, vitae viverra
          purus risus sed purus.
        </p>
      </ModalBody>
      <ModalFooter>
        <HFlow justifyContent='flex-end'>
          <ModalFooterButton onClick={action('cancel clicked')}>Cancel</ModalFooterButton>
          <ModalFooterButton
            kind='primary'
            onClick={modal({
              title: 'Confirm?',
              size: 'small',
              render: () => 'Confirm?',
              depthLevel: 2,
              manageOverflow: false,
              onClose: action('close'),
              actions: [
                { label: 'Cancel', onClick: action('Cancel') },
                { label: 'Ok', kind: 'primary', onClick: action('Ok') },
              ],
            })}
          >
            Open a modal that overlaps
          </ModalFooterButton>
        </HFlow>
      </ModalFooter>
    </Modal>
    <ModalMountTarget />
  </div>
)
