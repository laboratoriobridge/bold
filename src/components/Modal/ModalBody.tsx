import React, { useContext } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { ModalContext, ModalScroll } from './Modal'

export interface ModalBodyProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  style?: ExternalStyles
}

export function ModalBody(props: ModalBodyProps) {
  const { children, style, ...rest } = props

  const { scroll } = useContext(ModalContext)
  const { classes, css } = useStyles(createStyles, scroll)

  return (
    <div className={css(classes.body, style)} {...rest}>
      {children}
    </div>
  )
}

const createStyles = (theme: Theme, scroll: ModalScroll) => ({
  body: {
    overflow: scroll === 'paper' ? 'auto' : 'hidden',
    padding: '1rem 2rem 2rem 2rem',
    maxHeight: scroll === 'paper' ? 'calc(80vh - 5rem - 6rem)' : 'auto',
  } as React.CSSProperties,
})
