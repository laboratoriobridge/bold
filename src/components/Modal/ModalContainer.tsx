import React, { CSSProperties, forwardRef, useContext } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { ModalHeader, ModalHeaderProps } from './ModalHeader'
import { ModalContext, ModalScroll } from './Modal'

export interface ModalContainerProps
  extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style' | 'title'>,
    Omit<ModalHeaderProps, 'onCloseButtonClick'> {
  style?: ExternalStyles
  onClose?(): any
}

export const ModalContainer = forwardRef<HTMLDivElement, ModalContainerProps>((props, ref) => {
  const { style, onClose, hasCloseIcon, children, title, subtitle, icon, iconFill, ...rest } = props

  const { scroll } = useContext(ModalContext)
  const { classes, css } = useStyles(createStyles, scroll)

  return (
    <div role='dialog' aria-modal='true' ref={ref} className={css(classes.wrapper, style)} {...rest}>
      <ModalHeader
        title={title}
        subtitle={subtitle}
        icon={icon}
        iconFill={iconFill}
        hasCloseIcon={hasCloseIcon}
        onCloseButtonClick={onClose}
      />
      {children}
    </div>
  )
})

ModalContainer.defaultProps = {
  showCloseIcon: true,
  onClose: () => null,
} as Partial<ModalContainerProps>

const createStyles = (theme: Theme, scroll: ModalScroll) => ({
  wrapper: {
    maxHeight: '80vh',
    border: `1px solid ${theme.pallete.divider}`,
    boxShadow: theme.shadows.outer['160'],
    borderRadius: theme.radius.modal,
    backgroundColor: theme.pallete.surface.main,
    minWidth: 520,
    pointerEvents: 'auto',
    position: 'relative',
    overflow: scroll === 'body' ? 'hidden' : 'auto',
    display: 'grid',
    gridTemplateRows: scroll === 'body' ? '1fr auto auto' : 'initial',
  } as CSSProperties,
})
