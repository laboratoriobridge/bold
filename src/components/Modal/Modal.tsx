import FocusTrap from 'focus-trap-react'
import React, { useEffect, Ref } from 'react'
import { Theme, useStyles } from '../../styles'
import { zIndexLevel } from '../../styles/theme/zIndex'
import { Portal } from '../Portal'
import { FadeTransition } from '../Transition/FadeTransition'
import { ModalBackdrop } from './ModalBackdrop'
import { ModalContainer, ModalContainerProps } from './ModalContainer'

export type ModalSize = 'small' | 'large' | 'auto'
export type ModalDepthLevel = 1 | 2 | 3 | 4 | 5
export interface ModalProps extends ModalContainerProps {
  open: boolean
  size?: ModalSize
  children?: React.ReactNode
  containerRef?: Ref<HTMLDivElement>

  /**
   * @description allows you to customize the depth of the container and the backdrop of the modal
   * @default 1 - the lowest possible value
   */
  depthLevel?: ModalDepthLevel

  /**
   * @description allows you to remove the document's overflow property when a modal is closed
   * @default true
   */
  manageOverflow?: boolean

  /**
   * Specify whether the `onClose` prop should called when backdrop is clicked.
   * @default true
   */
  closeOnBackdropClick?: boolean
}

export function Modal(props: ModalProps) {
  const {
    open,
    size,
    closeOnBackdropClick,
    children,
    containerRef,
    style,
    onClose,
    depthLevel,
    manageOverflow,
    ...rest
  } = props
  const { classes, css } = useStyles(createStyles, depthLevel)

  // Kill body scroll when opened
  useEffect(() => {
    if (manageOverflow) {
      if (open) {
        document.body.classList.add(classes.bodyWhenOpened)
      } else {
        document.body.classList.remove(classes.bodyWhenOpened)
      }
      return () => document.body.classList.remove(classes.bodyWhenOpened)
    }
  }, [open, classes.bodyWhenOpened, manageOverflow])

  useEffect(() => {
    // Attach "Escape" to close modal
    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        onClose()
      }
    }

    if (open) {
      document.addEventListener('keydown', handleKeyDown)
    } else {
      document.removeEventListener('keydown', handleKeyDown)
    }
    return () => document.removeEventListener('keydown', handleKeyDown)
  }, [open, onClose])

  return (
    <FadeTransition in={open}>
      {({ className }) => (
        <>
          {open && (
            <Portal>
              <FocusTrap>
                <div className={className}>
                  <div className={classes.modal}>
                    <ModalContainer
                      ref={containerRef}
                      style={css(classes.container, classes[size], style)}
                      onClose={onClose}
                      {...rest}
                    >
                      {children}
                    </ModalContainer>
                  </div>

                  <ModalBackdrop depthLevel={depthLevel} onClick={closeOnBackdropClick ? onClose : undefined} />
                </div>
              </FocusTrap>
            </Portal>
          )}
        </>
      )}
    </FadeTransition>
  )
}

Modal.defaultProps = {
  size: 'large',
  closeOnBackdropClick: true,
  depthLevel: 1,
  manageOverflow: true,
} as Partial<ModalProps>

const createStyles = (theme: Theme, depthLevel: number) => ({
  modal: {
    position: 'fixed',
    left: '0',
    top: '0',
    width: '100%',
    height: '100vh',
    pointerEvents: 'none',
    zIndex: zIndexLevel[depthLevel].modalContainer,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    padding: '2rem',
  } as React.CSSProperties,
  container: {
    maxHeight: '80vh',
    overflow: 'auto',
  },
  bodyWhenOpened: {
    overflow: 'hidden',
  },
  large: { width: 900 },
  small: { width: 520 },
  auto: { maxWidth: '100%' },
})
