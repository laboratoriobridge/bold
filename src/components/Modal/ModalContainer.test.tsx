import { render } from '@testing-library/react'
import React from 'react'

import { ModalContextProps, ModalContextProvider } from '../../hooks/useModalContext'
import { ModalContainer } from './ModalContainer'
import { ModalHeader } from './ModalHeader'
import { ModalScroll } from './Modal'

jest.mock('../../util/string')

jest.mock('./ModalHeader', () => ({
  ModalHeader: jest.fn((props) => <div {...props} />),
}))

const mockContextValue: ModalContextProps = {
  bodyRef: { current: document.createElement('div') },
  scroll: 'body' as ModalScroll,
}

beforeEach(() => {
  jest.clearAllMocks()
})

it('should render correctly', () => {
  const { container } = render(
    <ModalContextProvider value={mockContextValue}>
      <ModalContainer title='Modal container'>Container</ModalContainer>
    </ModalContextProvider>
  )
  expect(container).toMatchSnapshot()
})

it('should accept the "style" prop', () => {
  const { container } = render(
    <ModalContextProvider value={mockContextValue}>
      <ModalContainer title='Modal container' style={{ color: 'red' }}>
        Container
      </ModalContainer>
    </ModalContextProvider>
  )
  expect(container).toMatchSnapshot()
})

it('should provide a ref to the div html element', () => {
  const ref = React.createRef<HTMLDivElement>()
  render(
    <ModalContextProvider value={mockContextValue}>
      <ModalContainer title='Modal container' ref={ref}>
        Container
      </ModalContainer>
    </ModalContextProvider>
  )
  expect(ref.current.tagName).toEqual('DIV')
})

describe('prop passing to ModalHeader', () => {
  it('passes all expected props correctly to ModalHeader', () => {
    const onClose = jest.fn()

    render(
      <ModalContextProvider value={mockContextValue}>
        <ModalContainer title='title' subtitle='subtitle' icon='infoCircleFilled' onClose={onClose} hasCloseIcon>
          Container
        </ModalContainer>
      </ModalContextProvider>
    )

    expect(ModalHeader).toHaveBeenCalledWith(
      expect.objectContaining({
        title: 'title',
        subtitle: 'subtitle',
        icon: 'infoCircleFilled',
        hasCloseIcon: true,
        onCloseButtonClick: onClose,
      }),
      {}
    )
  })
})
