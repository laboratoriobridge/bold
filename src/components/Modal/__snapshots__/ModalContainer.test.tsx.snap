// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept the "style" prop 1`] = `
.emotion-0 {
  max-height: 80vh;
  border: 1px solid #D3D4DD;
  box-shadow: 0 8px 10px -5px rgba(0,0,0,0.2),0 6px 30px 5px rgba(0,0,0,0.12),0 16px 24px 2px rgba(0,0,0,0.14);
  border-radius: 4px;
  background-color: #FFFFFF;
  min-width: 520px;
  pointer-events: auto;
  overflow: hidden;
  display: grid;
  grid-template-rows: 1fr auto;
  color: red;
}

<div>
  <div
    aria-modal="true"
    class="emotion-0"
    role="dialog"
    title="Modal container"
  >
    Container
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  max-height: 80vh;
  border: 1px solid #D3D4DD;
  box-shadow: 0 8px 10px -5px rgba(0,0,0,0.2),0 6px 30px 5px rgba(0,0,0,0.12),0 16px 24px 2px rgba(0,0,0,0.14);
  border-radius: 4px;
  background-color: #FFFFFF;
  min-width: 520px;
  pointer-events: auto;
  overflow: hidden;
  display: grid;
  grid-template-rows: 1fr auto;
}

<div>
  <div
    aria-modal="true"
    class="emotion-0"
    role="dialog"
    title="Modal container"
  >
    Container
  </div>
</div>
`;
